### ISSUE TITLE
Multiple AMIs design

### ISSUE THREAD

**Issue by danizen** [NLM-OCCS/occs-packer#11]
@reavon, how do you envision multiple AMIs being built?   To me, this is a question about the DRY (Don't Repeat Yourself) pattern; I care more about how much time it takes to write it than how much time it takes to run it.   So, my proposal is as follows:

* There are multiple ansible playbooks, e.g. `ansible/playbook/main.yml`, `ansible/playbook/ruby.yml`, etc.
* Which playbook to run becomes a packer variable that is passed to the `packer-build` script.
* The VM name and description (which become part of the AMI), are also packer variables.
* To build multiple AMIs, you run occs-packer multiple times with different arguments.

The advantage of this approach are that:
* There is less duplication of code than, for instance, having multiple template files.
* A change to the main template is reflected in all the AMIs.
* The playbooks, and the roles they include, become the way to have differences between the AMIs.

The disadvantage of this approach are that:
* It is less flexible, as only the ansible playbooks and roles can be different.
* It is easier to get the naming/description wrong.

What do you think?   We basically need to decide between a single template with variables and multiple templates with a bunch of duplication between them.

**Comment by danizen** [comment]
For instance, the only differences between a Ruby Application server and a base image are as follows:

```
$ diff CentOS-7-x86_64-Minimal.vbox.json CentOS-7-x86_64-Ruby.vbox.json
20c20
<     "vm_name": "centos7",
---
>     "vm_name": "centos7-ruby",
84c84
<       "playbook_file": "ansible/playbooks/base.yml",
---
>       "playbook_file": "ansible/playbooks/ruby.yml",
88c88,89
<         "ansible/roles/ansible-800-53"
---
>         "ansible/roles/ansible-800-53",
>         "ansible/roles/occsab-platform"
109c110
<          "description": "CentOS 7 Base Image",
---
>          "description": "CentOS 7 RubyApp Image",
```

**Comment by reavon** [comment]
What I had thought about is keeping just one .json file. In this file, we would have say the virtualbox-iso builder. Everything in that section can be variables in the top of the file. The only difference in each builder would be the name. Then we would have provisioner's setup to reference each individual builder name:

```json
  "provisioners": [
    {
      "type": "file",
      "only": [ "centos7-ami-base,centos7-ami-ruby,centos7-ami-python" ],
      "source": "files/tmp/CloudWatchMonitoringScripts-1.2.1.zip",
      "destination": "/tmp/CloudWatchMonitoringScripts-1.2.1.zip",
      "direction": "upload"
    },
    {
      "type": "ansible-local",
      "only": [ "centos7-ami-base" ],
      "playbook_file": "ansible/playbooks/base.yml",
      "role_paths": [
        "ansible/roles/ansible-cloud-init",
        "ansible/roles/ansible-directories",
        "ansible/roles/ansible-800-53"
      ]
    },
    {
      "type": "ansible-local",
      "only": [ "centos7-ami-ruby" ],
      "playbook_file": "ansible/playbooks/base.yml",
      "role_paths": [
        "ansible/roles/ansible-cloud-init",
        "ansible/roles/ansible-directories",
        "ansible/roles/ansible-800-53",
        "ansible/roles/occsab-platform",
        "ansible/roles/occsab-ruby"
      ]
    },
    {
      "type": "ansible-local",
      "only": [ "centos7-ami-python" ],
      "playbook_file": "ansible/playbooks/base.yml",
      "role_paths": [
        "ansible/roles/ansible-cloud-init",
        "ansible/roles/ansible-directories",
        "ansible/roles/ansible-800-53",
        "ansible/roles/occsab-platform",
        "ansible/roles/occsab-python"
      ]
    }
  ]
```

Keep in mind this is theory. 

You can read more about it in the section titled [Run On Specific Builds](https://www.packer.io/docs/templates/provisioners.html)

I had rather work on one json file and variable out as much as possible and make the differences in the images with the different provisioner sections.

**Comment by danizen** [comment]
Addressed with pull #16
